src/lib/sticky-header/sticky-header-dir.ts

```diff
@@ -0,0 +1,295 @@
+ */
+import {Component, Directive, Input, Output,
+  OnDestroy, AfterViewInit, ElementRef, Injectable, Optional} from '@angular/core';
+import {Scrollable} from '../core/overlay/scroll/scrollable';
+
+
+/**
+ * 'cdkStickyRegion' is a region contains one sticky header and other
+ * contents that user wants to put under the sticky-header. There only can
+ * be one sticky-header in one 'cdkStickyRegion'.
+ * If a user does not define a 'cdkStickyRegion' for a sticky-header, the direct
+ * parent node of the sticky-header will be set as the 'cdkStickyRegion'.
+ */
+@Directive({
+  selector: '[cdkStickyRegion]',
+})
+export class CdkStickyRegion {
+  constructor(private _element: ElementRef) { }
+
+  getElementRef(): ElementRef {
+    return this._element;
+  }
+}
+
+
+/**
+ * The 'cdkStickyHeader' is the header which user wants to be stuck at top of the
+ * scrollable container. The main logic is in the 'sticker()' function.
+ */
+const STICK_START_CLASS = 'stick-start';
+const STICK_END_CLASS = 'sticky-end';
+@Directive({
+  selector: '[cdkStickyHeader]',
+})
+export class CdkStickyHeader implements OnDestroy, AfterViewInit {
+
+  /**
+   * Set the sticky-header's z-index as 10 in default. Make it as an input
+   * variable to make user be able to customize the zIndex when
+   * the sticky-header's zIndex is not the largest in current page.
+   * Because if the sticky-header's zIndex is not the largest in current page,
+   * it may be sheltered by other element when being stick.
+   */
+  @Input('cdkStickyHeaderZIndex') zIndex: number = 10;
+  @Input('cdkStickyParentRegion') parentRegion: HTMLElement;
+  @Input('cdkStickyScrollableRegion') scrollableRegion: HTMLElement;
+
+  private _onScrollBind: EventListener = this.onScroll.bind(this);
+  private _onResizeBind: EventListener = this.onResize.bind(this);
+  private _onTouchMoveBind: EventListener = this.onTouchMove.bind(this);
+  public isStuck: boolean = false;
+
+  // the element with the 'cdkStickyHeader' tag
+  public element: HTMLElement;
+
+  // the upper container element with the 'cdkStickyRegion' tag
+  public stickyParent: HTMLElement;
+  public upperScrollableContainer: HTMLElement;
+
+  /**
+   * the original css of the sticky element, used to reset the sticky element
+   * when it is being unstuck
+   */
+  public originalCss: any;
+
+  private _containerStart: number;
+  private _scrollFinish: number;
+
+  private _scrollingWidth: number;
+
+  constructor(private _element: ElementRef,
+              public scrollable: Scrollable,
+              @Optional() public parentReg: CdkStickyRegion) {
+    this.element = _element.nativeElement;
+    this.upperScrollableContainer = scrollable.getElementRef().nativeElement;
+    this.scrollableRegion = scrollable.getElementRef().nativeElement;
+    if (parentReg != null) {
+      this.parentRegion = parentReg.getElementRef().nativeElement;
+    }
+  }
+
+  ngAfterViewInit(): void {
+
+    if (this.parentRegion != null) {
```

Comment:
`this.stickyParent = this.parentRegion != null ? this.parentRegion : this.element.parentElement`